pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1042 @@
+	///   the states accordingly.
+	#[allow(clippy::type_complexity)]
+	fn handle_concurrent_swap_orders(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		swap_order_id: T::TokenSwapOrderId,
+		reason: TokenSwapReason,
+	) -> Result<
+		(
+			Option<Swap<T::Balance, T::CurrencyId>>,
+			Option<InvestState<T::Balance, T::CurrencyId>>,
+			Option<RedeemState<T::Balance, T::CurrencyId>>,
+		),
+		DispatchError,
+	> {
+		let last_reason = ForeignInvestmentInfo::<T>::get(swap_order_id)
+			.ok_or(Error::<T>::InvestmentInfoNotFound)?
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		// Exit early if both reasons match, i.e. we would not override any opposite
+		// swap order
+		if last_reason == reason {
+			return Ok((None, None, None));
+		}
+
+		// Read states from storage and determine amounts
+		let invest_state = InvestmentState::<T>::get(who, investment_id).unwrap_or_default();
+		let redeem_state = RedemptionState::<T>::get(who, investment_id).unwrap_or_default();
+		let invest_swap_amount = invest_state
+			.get_active_swap()
+			.map(|s| s.amount)
+			.unwrap_or_default();
+		let redeem_swap_amount = redeem_state
+			.get_active_swap()
+			.map(|s| s.amount)
+			.unwrap_or_default();
+		let resolved_amount = invest_swap_amount.min(redeem_swap_amount);
+		// safe because max >= min, equals zero if both amounts equal
+		let swap_amount_opposite_direction =
+			invest_swap_amount.max(redeem_swap_amount) - resolved_amount;
+
+		// Determine new invest state
+		let new_invest_state = match invest_state {
+			// As redeem swap can only be into return currency, we need to delta on the opposite
+			// swap directions
+			InvestState::ActiveSwapIntoPoolCurrency { swap } => {
+				if invest_swap_amount > redeem_swap_amount {
+					Ok(Some(
+						InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+							swap: Swap {
+								amount: swap_amount_opposite_direction,
+								..swap
+							},
+							invest_amount: resolved_amount,
+						},
+					))
+				} else {
+					Ok(Some(InvestState::InvestmentOngoing {
+						invest_amount: resolved_amount,
+					}))
+				}
+			}
+			// Same as above except for the base investment amount which is incremented
+			InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+				swap: invest_swap,
+				invest_amount,
+			} => {
+				if invest_swap_amount > redeem_swap_amount {
+					Ok(Some(
+						InvestState::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+							swap: Swap {
+								amount: swap_amount_opposite_direction,
+								..invest_swap
+							},
+							invest_amount: invest_amount.ensure_add(resolved_amount)?,
+						},
+					))
+				} else {
+					Ok(Some(InvestState::InvestmentOngoing {
+						invest_amount: invest_amount.ensure_add(resolved_amount)?,
+					}))
+				}
+			}
+			// We must not alter the invest state if there is no active pool currency swap
+			_ => Ok(None),
+		}
+		.map_err(|e: DispatchError| e)?;
```

Comment:
I think there is no need to return this variable as a `Result`.

Reply:
Great catch, you are right!